### ISSUE TITLE
Is there a way to extend the Redis InvisibilityTimeout within a given job?

### ISSUE THREAD

**Issue by dcorbett-fivebridgesllc** [HangfireIO/Hangfire#936]
I have jobs that can run in seconds, and other jobs that run in hours.   It would be very useful to be able to extend the Invisibility Timeout in Redis programmatically.

Is there any way this can be done?

I'd be willing to submit a PR for this, if it requires a modification to Hangfire.

Thanks,

    - Daniel

**Comment by odinserj** [comment]
I have an idea to get rid of Invisibility Timeout, replacing it by Sliding Invisibility Timeout technique as in the [recent](https://www.hangfire.io/blog/2017/06/16/hangfire-1.6.14.html) Hangfire.SqlServer release. The idea is to have a short-lived timeout (1 or 5 minutes), but increase it using a timer while processing is taking place. But I want to release Hangfire 1.7.0 first before continuing to work on Hangfire.Pro.Redis.

**Comment by danielcor** [comment]
Thanks!  This will help us.  With this technique, how do you handle a
particular job that is deadlocked or hung?   How do you tell me the
difference between it and something that is merely long running?

   - Daniel

**Comment by odinserj** [comment]
Hangfire does not handle deadlocked or hanging jobs – it's up to developers who are using Hangfire to eliminate these kinds of issues in application code that is considered as background jobs with help of timeouts, etc.

Sliding invisibility timeout simply helps workers to fetch terminated background jobs with fewer delay, when host process was non-gracefully terminated.

**Comment by danielcor** [comment]
Thanks.   Makes sense.   So this is solely about ungraceful termination of
Hangfire and how we handle restarting jobs.

   - Daniel

**Comment by odinserj** [comment]
Yes, invisibility is all about hiding a fetched background job from other workers. Timeout is here to ensure that in case of a process termination, background job id will eventually be visible for other workers, so they can proceed with the processing. Under normal circumstances, and if all the shutdowns are graceful (`BackgroundJobServer.Dispose` method is called and awaited), background job id is re-queued immediately.

**Comment by Elbert-Oh** [comment]
It would really be nice if this was explained properly in the documentation. It took me forever to find out what this option (**SlidingInvisibilityTimeout**) does.
